File: one-unity/core/development/common/game-account/Runtime/Scripts/Callback/AuthResult.cs
namespace TPFive.Game.Account
{
    public class AuthResult<T>
    {
        public bool IsSuccess => Error == null && Payload != null;

        public T Payload { get; set; }

        public AuthError Error { get; set; }

        public static AuthResult<T> Success(T payload)
        {
            return new AuthResult<T>()
            {
                Payload = payload,
            };
        }

        public static AuthResult<T> Failure(AuthError error)
        {
            return new AuthResult<T>()
            {
                Error = error,
            };
        }

        public static AuthResult<T> Failure(int code, string message)
        {
            return new AuthResult<T>()
            {
                Error = new AuthError(code, message),
            };
        }
    }
}

File: one-mobile/flutter_project/lib/xrauth/callback/auth_result.dart
import 'package:tpfive/xrauth/constants/auth_error.dart';

class AuthResult<T> {
  bool get isSuccess => error == null && payload != null;

  T? payload;

  AuthError? error;

  AuthResult({this.payload, this.error});

  @override
  String toString() {
    return 'AuthResult{error: $error}';
  }

  static AuthResult<T> success<T>(T payload) {
    return AuthResult<T>(payload: payload);
  }

  static AuthResult<T> failure<T>(AuthError error) {
    return AuthResult<T>(error: error);
  }

  static AuthResult<T> failureWithCode<T>(int code, String message) {
    return AuthResult<T>(error: AuthError(code, message));
  }
}
